externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/FlinkEngineUtils.scala

```diff
@@ -62,47 +70,105 @@ object FlinkEngineUtils extends Logging {
+        if (!dir.canRead) throw new SqlClientException(s"Directory cannot be read: $dir")
+        val files: Array[File] = dir.listFiles
+        if (files == null) throw new SqlClientException(s"Directory cannot be read: $dir")
+        for (f <- files.filter(f => f.isFile && f.getAbsolutePath.toLowerCase.endsWith(".jar"))) {
+          val url: URL = f.toURI.toURL
+          JarUtils.checkJarFile(url)
+          dependencies.add(url)
+        }
+      }
+    } catch {
+      case e: Exception =>
+        throw new SqlClientException("Could not load all required JAR files.", e)
+    }
+    dependencies
+  }
+
+  def getDefaultContext(
+      args: Array[String],
+      flinkConf: Configuration,
+      flinkConfDir: String): DefaultContext = {
+    val parser = new DefaultParser
+    val line = parser.parse(EMBEDDED_MODE_CLIENT_OPTIONS, args, true)
+    val jars: JList[URL] = Option(checkUrls(line, CliOptionsParser.OPTION_JAR))
+      .getOrElse(JCollections.emptyList())
+    val libDirs: JList[URL] = Option(checkUrls(line, CliOptionsParser.OPTION_LIBRARY))
+      .getOrElse(JCollections.emptyList())
+    val dependencies: JList[URL] = discoverDependencies(jars, libDirs)
+    if (FlinkEngineUtils.isFlinkVersionEqualTo("1.16")) {
+      val commandLines: JList[CustomCommandLine] =
+        Seq(new GenericCLI(flinkConf, flinkConfDir), new DefaultCLI).asJava
+      DynConstructors.builder()
+        .impl(
+          classOf[DefaultContext],
+          classOf[Configuration],
+          classOf[JList[CustomCommandLine]])
+        .build()
+        .newInstance(flinkConf, commandLines)
+        .asInstanceOf[DefaultContext]
+    } else if (FlinkEngineUtils.isFlinkVersionEqualTo("1.17")) {
+      DynMethods.builder("load")
+        .impl(
+          classOf[DefaultContext],
+          classOf[Configuration],
+          classOf[JList[URL]],
+          classOf[Boolean],
+          classOf[Boolean])
+        .buildStatic()
+        .invoke[DefaultContext](
+          flinkConf,
+          dependencies,
+          new JBoolean(true),
+          new JBoolean(false))
     } else {
-      throw new SqlClientException("Other mode is not supported yet.")
+      throw new KyuubiException(
+        s"Flink version ${EnvironmentInformation.getVersion} are not supported currently.")
     }
   }
 
-  def getSessionContext(localExecutor: LocalExecutor, sessionId: String): SessionContext = {
-    val method = classOf[LocalExecutor].getDeclaredMethod("getSessionContext", classOf[String])
-    method.setAccessible(true)
-    method.invoke(localExecutor, sessionId).asInstanceOf[SessionContext]
+  def getSessionContext(session: Session): SessionContext = {
+    DynFields.builder()
+      .hiddenImpl(classOf[Session], "sessionContext")
+      .build()
+      .get(session)
+      .asInstanceOf[SessionContext]
   }
 
-  def parseEmbeddedModeClient(args: Array[String]): CliOptions =
+  def getResultJobId(resultFetch: ResultFetcher): JobID = {
```

Comment:
```suggestion
  def getResultJobId(resultFetch: ResultFetcher): Option[JobID] = {
```

Reply:
Generally, we should use `Option` to represent nullable value as much as possible